update ScreenSelectMusic's SortMenu UIUX and data structure

quietly-turning opened the pull request:
## UIUX Changes
This changes the UI for ScreenSelectMusic's SortMenu's choices (like "sort by title", "sort by artist", "change style to double") visually distinct from folders-of-choices (like "Sorts", "Options").

Folders-of-choices now:
  * have a background quad
  * have left-aligned text
  * include a folder icon

The [recent change](https://github.com/Simply-Love/Simply-Love-SM5/pull/585) that allowed the SortMenu to use a folder/nested-choice structure was a good change(!), though I sometimes found myself getting visually lost in the menu.  My primary goal with this PR was to make folder-rows more visually distinct from choice-rows.  

The folder icon comes from https://github.com/tailwindlabs/heroicons, which is [MIT licenced](https://github.com/tailwindlabs/heroicons?tab=MIT-1-ov-file#readme).

This also changes the UX of SSM's SortMenu folder to behave similar to the engine's MusicWheel: toggling a folder open reveals new rows below it, and opening one new folder auto-closes the previously-open folder.

https://github.com/user-attachments/assets/2bf7636d-48d0-4f65-9a47-bd0876fc85c7


---

## Code Changes

On the code side of things, there's 3 main changes:


1.  I moved a lot of code out of default.lua.  My motivation was to be looking at fewer lines-of-code at any given moment.
  * moved InputHandlers into a discrete directory for InputHandlers
  * moved all SongSearch code into the existing SongSearch directory
  * moved `wheel_choices` into discrete file

2. I restructured SortMenu's primary `wheel_choices` data structure to an array of keyed tables, each representing one folder:
 * `name` (string)
 * `open` (boolean)
 * `children` (table).

The `children` table is an array with each item in the format of 
```
{{ top_text, bottom_text, action_if_chosen}, condition_to_be_visible }
```

My motivation was to have a data structure with a more-clear distinction between a folder-row and a choice-row.

You can see this new structure here: https://github.com/Simply-Love/Simply-Love-SM5/blob/e0110ca6500240c2a865e3859ddefa192f03f02b/BGAnimations/ScreenSelectMusic%20overlay/SortMenu/SortMenuRows.lua#L4-L84

3. I moved "action" code (what happens when the user presses START on a given choice in the SortMenu) out of SortMenu_InputHandler, into a dedicated file for reusable helper functions, and those helper functions are referenced as the 3rd element in each child row (denoted above as `action_if_chosen`).

The SortMenu_InputHandler was previously making some complicated inferences to determine what kind of row it was operating on, like knowing the current row was a folder because the top_text started with the substring "Category", or knowing the current row was a playlist because it had a `new_overlay` key.  It worked, but the [long if-else chain](https://github.com/Simply-Love/Simply-Love-SM5/blob/ed98aad9934176c85e8052e238cb04ad09e475d9/BGAnimations/ScreenSelectMusic%20overlay/SortMenu/SortMenu_InputHandler.lua#L18-L150) could be tricky for an unfamiliar dev to work with.

I think the [new structure](https://github.com/Simply-Love/Simply-Love-SM5/blob/e0110ca6500240c2a865e3859ddefa192f03f02b/BGAnimations/ScreenSelectMusic%20overlay/SortMenu/SortMenuRows.lua#L4-L84) should be easier for future devs to jump into — the function to dictate what happens when a row is chosen is bundled with that row.

---

In all, I changed a lot of code, so feel free to ask questions!  I did try to write lots of new inline comments where I thought they'd help.

It'd be good if @CrashCringle12 could test this branch and both ensure I didn't break existing functionality and give feedback.

quietly-turning committed:
overhaul SSM's SortMenu

This changes the UI for ScreenSelectMusic's SortMenu's choices (like
"sort by title", "sort by artist", "change style to double") visually
distinct from folders-of-choices (like "Sorts", "Options").

Folders-of-choices now:
  * have a background quad
  * have left-aligned text
  * include a folder icon

This also changes the UX of SSM's SortMenu folder to behave similar to
the engine's MusicWheel: toggling a folder open reveals new rows below
it, and opening one new folder auto-closes the previously-open folder.
The previous UX was to fully drill-down-into a folder when opening it,
replacing all rows in the SortMenu with only the choices available in the new
folder.

---

On the code side of things, this moves code out of default.lua:
  * moves InputHandlers into a discrete directory for InputHandlers
  * moves all SongSearch code into the existing SongSearch directory

It also restructures SortMenu's primary `wheel_choices` data structure
to use key/value pairs: name (string), open (boolean), children (table).

The `children` table is an array with each item in the format of
{{ top_text, bottom_text, action_if_chosen()}, condition_to_be_visible }

It also moves "action" code (what happens when the user presses START on
a given choice in the SortMenu) out of SortMenu_InputHandler, and into a
dedicated file for reusable helper functions, and those helper functions
are referenced as the 3rd element in each child row (denoted above as
action_if_chosen).

The SortMenu_InputHandler was previously making some complicated
inferences to determine what kind of row it was operating on, like
knowing the current row was a folder because the top_text started with
the substring "Category", or knowing the current row was a playlist
because it had a `new_overlay` key.

It worked, but I think the new structure is cleaner and hopefully easier
to future devs to jump into -- the function to dictate what happens when
a row is chosen is bundled with that row.

quietly-turning commented on BGAnimations/ScreenSelectMusic overlay/SortMenu/SortMenuRows.lua:
```diff
@@ -0,0 +1,86 @@
+local  ShowSongSearch, ShowTestInput, ShowLeaderboard, ShowDownloads, ShowPracticeMode, ShowSelectProfile, ShowSetSummary, ShowLoadNewSongs, ChangeSort, ChangeMode, ChangeStyle, AddSongToFavorites, AddFavoritesRow, AddPlaylistsRows, GetChangeableStylesRows, DownloadsExist = unpack(LoadActor("./SortMenuHelpers.lua", ...))
```
Is there a better way to do an "import" like this in StepMania-flavored Lua?

quietly-turning commented:
Give me, like, one more day to add to language files to the best of my ability and to clean up one more piece of code.

CrashCringle12 committed:
Move the wheel_options into the SortMenu actor so it's accessible from other areas

quietly-turning commented:
> Give me, like, one more day to add to language files to the best of my ability and to clean up one more piece of code.

Okay, I'm good now.  :)

I included an additional commit to incorporate @CrashCringle12's proposed changes from https://github.com/Simply-Love/Simply-Love-SM5/pull/665 that exposed the SortMenu's `wheel_options` to SL modules.

quietly-turning commented:
@SimplyViper16 @JustMoneko @TheBreak1 
Sorry to ping you! 😅 

Can you help proofread the machine translation I used in this PR?  Thank you for considering!

SimplyViper16 committed:
update IT language file

JustMoneko committed:
Update pl.ini

teejusb commented:
Just to confirm -- is https://github.com/Simply-Love/Simply-Love-SM5/pull/665 still required or would this PR cover it?

quietly-turning commented:
Thanks for the thorough review, @CrashCringle12!  

The bugs you found could be fixed, but you've given me a lot to think about (e.g. choices that affect both players vs. choices that affect one player) and I'm no longer confident this PR is a good change I want to work more on.

I'll reflect on it for a week or so and respond here.

quietly-turning committed:
ensure SortMenu strings exist before setting them

With user-supplied modules editing the contents of SSM's SortMenu, it's
not safe to assume strings exist in SL language files.

We can use THEME:HasString() to check if a string exists in the language
file for the current lanague, localize it using THEME:GetString() if so
and fall back on using the literal string supplied to SortMenuRows if
not.

quietly-turning committed:
change gamestate from SortMenu more defensively

Since user-supplied modules and curious end-users can modify the
contents of SSM's SortMenu, it's not safe to assume all choices will be
valid.

This adds validation to SSM's SortMenu for
  * ChangeSort()
  * ChangeStyle()
  * ChangeMode()

helper functions.  If an invalid Sort, Style, or Mode is supplied to
these helpers, warn the user usig ITGm's lua.ReportScriptError() and
play SL's "common invalid" sound effect.

quietly-turning committed:
adjust SSM's SortMenu layout

*  make the SortMenu slightly taller to accommodate:
   1 item with focus in the middle
   2 full items above, 2 full items below
   1 half-item above, 1 half-item above

* change SortMenu's masks and bottom-help-text to set y-offet using
  SortMenu's dimensions, rather than hardcoded magic numbers

* add inline comments noting where responsive layout (i.e. if the
  SortMenu grows taller or shorter in the future) could be improved but
  are outside the scope of this current PR effort

quietly-turning committed:
don't try to use SortMenu's toptext as a flag

When the SortMenu builds its simple array of WheelItems to prsent to the
player, I'd previously relied the first index of each WheelItem as both
top_text (like "Change Sort To" and "Feeling Salty?") *and* a flag to
determine if this WheelItem represented a folder or a choice.  I was
using code like

if (info[1] == "ToggleFolder") then

where top_text "ToggleFolder" was a special value.  That worked so long
as we could ensure every top_text value existed in SL's language files:
  it's a choice if the string exists, and
  it's a folder if the string doesn't exist, because there was no
    "ToggleFolder" in en.ini

Since SortMenu modules need to add their own folders and choices, they
can't rely on text that exists in SL's language files; they'll more than
likely need to provide strings for top_text and bottom_text as-is,
without any THEME:GetString() localization.

That means we cannot rely on (info[1]=="ToggleFodler") as a flag.
(Probably never should have, but hey.)

WheelItems are now shaped like:
`{ top_text, bottom_text, action_if_chosen, is_folder}`

`top_text` and `bottom_text` are strings
`action_if_chosen` is a reference to a function, and
`is_folder` is a boolean

SL modules should be able to add folders and choices to SSM's SortMenu
like

`
t["ScreenSelectMusic"] = Def.ActorFrame {
  ModuleCommand=function(self)
    local sortmenu = SCREENMAN:GetTopScreen():GetChild("Overlay"):GetChild("SortMenu")
    table.insert(sortmenu.wheel_options,
      {
        name="Profile Sorts",
        open=false,
        children={
          { {"Show", "My Recently Played", ProfileSortRecent},     atLeastOneProfile },
          { {"Show", "My High Scores",     ProfileSortTopGrades},  atLeastOneProfile },
          { {"Show", "My Most Played",     ProfileSortPopularity}, atLeastOneProfile },
        }
      }
    )
  end
}
`

In this^ example, text like "Show" and "My High Scores" will be
presented as-is, no localization strings from SL's language files
necessary.

quietly-turning committed:
let SortMenuHelper functions return false on error

SortMenu_InputHandler calls the action_if_chosen functions for each
WheelItem -- we may as well have those functions return false if they
hit an error condition so that SortMenu_InputHandler can respond (i.e.
play the "common invalid" sound effect).

quietly-turning committed:
only show one "Favorites" SortMenu WheelItem

I'd previously added separate SortMenu WheelItems for "P1 Favorites" and
"P2 Favorites".  CrashCringle suggested condensing those to a single
"Favorites" and letting MenuButton input dictate which favorites the
MusicWheel then showed.

quietly-turning committed:
evalaute children conditions before adding folder

For SSM's SortMenu, a folder's children can have conditions that are

* absent (always add this row to the Sort Menu)
* an expression that evaluates to boolean (evaluate it once at screen
  init)
* a function that returns a boolean (evaluate it each time
  AssessAvailableChoicesCommand is called)

A folder could have children whose conditions are all functions, which
could all evaluate to false, which means we should evaluate those
functions before adding a WheelItem for the folder, or else we could get
a folder with 0 children.

quietly-turning committed:
reset SortMenu focus after SwitchProfile

quietly-turning commented:
> I plan to work on this a little more when I have time, either late December 2025 or January 2026.

I didn't make time to work on this, and am preparing for my step-back-from-all-responsibilities-once-every-five-years-sabbatical, so I'll close this PR now.

FWIW, I think as-is this PR offers modest UIUX improvement:
* folders are visually-distinct from choices
* the wheel maintains state of the most-recently-selected choice and will re-open there if possible

and I don't know of any outstanding bugs with the implementation.

I'm in favor it being merged, but don't plan to work on it any further at this time.
